cardano-api/src/Cardano/Api/Value.hs

```diff
@@ -115,6 +131,98 @@ instance Semigroup Value where
 instance Monoid Value where
   mempty = Value Map.empty
 
+instance ToJSON Value where
+  toJSON = toJSON . unflatten
+
+unflatten :: Value -> AssetsBundle
+unflatten (Value flatMap) = AssetsBundle $ Map.foldlWithKey folder [] flatMap
+ where
+  folder
+    :: [AssetIdBundle]
+    -> AssetId
+    -> Quantity
+    -> [AssetIdBundle]
+  folder acc (AssetId pid aName) q = AssetIdBundle pid (Map.singleton aName q) : acc
+  folder acc AdaAssetId q = AdaAsset q : acc
+
+instance FromJSON Value where
+  parseJSON jv = flatten <$> parseJSON jv
+
+flatten :: AssetsBundle -> Value
+flatten (AssetsBundle bundles) = foldl folder (Value Map.empty) bundles
+ where
+  folder :: Value -> AssetIdBundle -> Value
+  folder acc (AssetIdBundle pid assetNameQuantityMap) =
+   acc <> Value (Map.mapKeys (\aname -> AssetId pid aname) assetNameQuantityMap)
+  folder acc (AdaAsset q) = acc <> Value (Map.singleton AdaAssetId q)
```

Comment:
I think this can be done in a "bulk" style rather than a foldl style, and would be clearer for it.

e.g. a `Map.fromList` on an appropriate list comprehension.